core/src/main/java/com/linecorp/armeria/client/pool/DefaultKeyedChannelPool.java

```diff
@@ -280,6 +287,12 @@ protected boolean offerChannel(K key, Channel channel) {
 
     @Override
     public void close() {
+        for (Channel ch : createdChannels) {
+            if (ch.isOpen()) {
+                ch.close().syncUninterruptibly();
```

Comment:
Actually, this could be done asynchronously and you can wait for the future after calling close() for all channels.

Reply:
I thought about this more and realized even with the current code, I think it would deadlock if close is called from the pool's event loop thread since the callbacks would need the event loop to be free to run. I can't think of a way to fix this and keep the synchronous nature of `close()` - I could make `close()` return a promise, but then it wouldn't implement `AutoCloseable` anymore. 

I'm thinking of just leaving these as `sync` since I guess it's not a huge deal if `close()` doesn't block - what do you think?

Reply:
I'm fine with closing synchronously, but we can make the process faster like this:

```java
Queue<ChannelFuture> closeFutures = new ArrayDeque<>();
for (Channel ch : allChannels) {
    closeFutures.add(ch.close());
}

for (;;) {
    ChannelFuture f = closeFutures.poll();
    if (f == null) break;
    f.syncUninterruptibly();
}
```

Reply:
Yeah but now I'm worried that that will deadlock if `close` is called from the event loop. Do you think it will work? Technically, it looks like it wouldn't lock and throw an exception instead

```java
protected void checkDeadLock() {
        EventExecutor e = executor();
        if (e != null && e.inEventLoop()) {
            throw new BlockingOperationException(toString());
        }
    }
```

I could make it synchronous when off event loop and async otherwise, but seems a little weird.

Reply:
> I could make it synchronous when off event loop and async otherwise, but seems a little weird.

Yeah, a little bit weird, but that's probably the best I can.